repose-aggregator/components/filters/cors-filter/src/main/scala/org/openrepose/filters/cors/CorsFilter.scala

```diff
@@ -178,23 +204,71 @@ class CorsFilter @Inject()(configurationService: ConfigurationService)
         Nil
     })
   }
+
+  def getHostUri(request: HttpServletRequestWrapper): URI = {
+    val maybeForwardedHost = request.getSplittableHeaderScala(CommonHttpHeader.X_FORWARDED_HOST).headOption
+    val maybeForwardedHostUri = Try(maybeForwardedHost
+      .map(forwardedHost => new URIBuilder(s"${request.getScheme}://$forwardedHost"))
+      .map(uri => uri.setPort(normalizePort(uri.getPort, uri.getScheme)).setHost(normalizeUriHost(uri.getHost)).build()))
+    lazy val hostUri = new URIBuilder().setScheme(request.getScheme).setHost(normalizeHost(request.getServerName))
+      .setPort(normalizePort(request.getServerPort, request.getScheme)).build()
+    maybeForwardedHostUri match {
+      case Success(None) =>
+        // there was no X-Forwarded-Host header in the request; don't bother logging this since this is the standard case
+        hostUri
+      case Success(Some(forwardedHostUri)) =>
+        logger.trace("Using X-Forwarded-Host header '{}' for host.", maybeForwardedHost.get)
+        forwardedHostUri
+      case Failure(exception) =>
+        logger.trace("Unable to parse X-Forwarded-Host header '{}'. Defaulting to Host header.", maybeForwardedHost.get, exception)
+        hostUri
+    }
+  }
+
+  def getOriginUri(origin: String): Try[URI] = Try(new URIBuilder(origin))
+    .map(originUri => originUri
+      .setPort(normalizePort(originUri.getPort, originUri.getScheme))
+      .setHost(normalizeUriHost(originUri.getHost))
+      .build())
+
+  def normalizeHost(host: String): String =
+    if (InetAddresses.isInetAddress(host)) InetAddresses.forString(host).getHostAddress else host
+
+  def normalizeUriHost(host: String): String =
+    if (InetAddresses.isUriInetAddress(host)) InetAddresses.forUriString(host).getHostAddress else host
+
+  def normalizePort(port: Int, scheme: String): Int = (port, scheme.toLowerCase) match {
+    case (p, _) if p > 0 => p
+    case (_, s) if s == "http" => 80
+    case (_, s) if s == "https" => 443
+    case _ => port
+  }
+
+  def getHeaderNamesToExpose(response: HttpServletResponseWrapper): List[String] =
+    (DefaultExposeHeaders.filterNot(response.containsHeader) ++ response.getHeaderNames.asScala).distinct
 }
 
 object CorsFilter {
-  private final val DEFAULT_CONFIG = "cors.cfg.xml"
-  private final val SCHEMA_FILE_NAME = "/META-INF/schema/config/cors-configuration.xsd"
+  private final val DefaultConfig = "cors.cfg.xml"
+  private final val SchemaFilename = "/META-INF/schema/config/cors-configuration.xsd"
+  private final val DefaultExposeHeaders = List(CommonHttpHeader.CONTENT_LENGTH.toString)
 
   implicit def autoHeaderToString(hc: HeaderConstant): String = hc.toString
 
-  sealed trait RequestType
-  object NonCorsRequest extends RequestType
-  object PreflightRequest extends RequestType
-  object ActualRequest extends RequestType
+  sealed trait RequestType {
+    def isOptions: Boolean
```

Comment:
Also, traits can totally hold 'val's. That may have been more appropriate for what you were trying to express -- namely, that every instance of this trait should have a readable `isOptions` value. Granted, the generated bytecode will probably look very similar if not identical.